### PR TITLE
Add some of the io for the field's metadata

### PR DESCRIPTION
**Pull request by uramirez8707**
**Description**

- Writes the field's metadata to the corresponding netcdf file. There is still more work needed for the diurnal variables and the uncompressed land files, but that will be done in a seperate PR

Fixes # (issue)

**How Has This Been Tested?**
CI

**Checklist:**
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules
- [x] New check tests, if applicable, are included
- [x] `make distcheck` passes

### CODE REVIEW

File: diag_manager/fms_diag_object.F90 [hunk 1]
```diff
@@ -428,18 +430,35 @@ end function fms_diag_axis_init
...
 
+  call this%fms_diag_do_io(time_step)
+
+#endif
+
+end subroutine fms_diag_send_complete
+
+!> @brief Loops through all the files, open the file, writes out axis and
+!! variable metadata and data when necessary.
+subroutine fms_diag_do_io(this, time_step, is_end_of_run)
```
[the remark this answers is not in the document]
Comment (thomas-robinson) [reply diag_manager/fms_diag_object.F90:448]: OK, we should also reach out to the scientists and make sure that they are ok with this behavior change

File: diag_manager/fms_diag_file_object.F90 [hunk 2]
```diff
@@ -908,21 +953,21 @@ subroutine write_time_metadata(this)
 
   if (diag_file%time_ops) then
     call register_variable_attribute(fileobj, time_var_name, "bounds", &
-      trim(time_var_name)//"_bounds", str_len=len_trim(time_var_name//"_bounds"))
+      trim(time_var_name)//"_bnds", str_len=len_trim(time_var_name//"_bnds"))
```
Comment (thomas-robinson) [inline diag_manager/fms_diag_file_object.F90:956]: Is this to get it in like with the current?

File: diag_manager/fms_diag_file_object.F90 [hunk 3]
```diff
@@ -1096,6 +1141,46 @@ subroutine write_axis_metadata(this, diag_axis)
...
+  integer :: j           !< For do loops
+  logical :: is_regional !< Flag indicating if the field is in a regional file
+  character(len=255) :: cell_measures
+
+  is_regional = this%is_regional()
+
+  diag_file => this%FMS_diag_file
+  fileobj => diag_file%fileobj
+
+  do i = 1, size(diag_file%field_ids)
```
Comment (thomas-robinson) [inline diag_manager/fms_diag_file_object.F90:1163]: Should/could this be a `do concurrent`?

File: diag_manager/fms_diag_file_object.F90 [hunk 4]
```diff
@@ -1096,6 +1141,46 @@ subroutine write_axis_metadata(this, diag_axis)
...
+subroutine write_field_metadata(this, diag_field, diag_axis)
+  class(fmsDiagFileContainer_type), intent(inout), target :: this            !< The file object
+  class(fmsDiagField_type)        , intent(inout), target :: diag_field(:)   !<
+  class(fmsDiagAxisContainer_type), intent(in)            :: diag_axis(:)    !< Diag_axis object
+
+  class(FmsNetcdfFile_t),  pointer     :: fileobj        !< The fileobj to write to
+  class(fmsDiagFile_type), pointer     :: diag_file      !< Diag_file object to open
+
+  integer :: i           !< For do loops
+  integer :: j           !< For do loops
```
Comment (thomas-robinson) [inline diag_manager/fms_diag_file_object.F90:1154]: This is a field ID not an index for looping.  It should probably be a pointer.

File: diag_manager/fms_diag_file_object.F90 [hunk 5]
```diff
@@ -1096,6 +1141,46 @@ subroutine write_axis_metadata(this, diag_axis)
...
+
+  diag_file => this%FMS_diag_file
+  fileobj => diag_file%fileobj
+
+  do i = 1, size(diag_file%field_ids)
+    j = diag_file%field_ids(i)
+    if (.not. diag_file%field_registered(i)) cycle !TODO do something else here
+
+    !TODO I think if the area and the volume field are no in the same file, a global attribute containing the
+    !the file that the fields are in needs to be added
```
Comment (thomas-robinson) [inline diag_manager/fms_diag_file_object.F90:1168]: Do we need to discuss this?

File: diag_manager/fms_diag_file_object.F90 [hunk 6]
```diff
@@ -1096,6 +1141,46 @@ subroutine write_axis_metadata(this, diag_axis)
...
+  character(len=255) :: cell_measures
+
+  is_regional = this%is_regional()
+
+  diag_file => this%FMS_diag_file
+  fileobj => diag_file%fileobj
+
+  do i = 1, size(diag_file%field_ids)
+    j = diag_file%field_ids(i)
+    if (.not. diag_file%field_registered(i)) cycle !TODO do something else here
```
Comment (thomas-robinson) [inline diag_manager/fms_diag_file_object.F90:1165]: Maybe instead of a `do` this could be a `where`

**Review by thomas-robinson — CHANGES_REQUESTED** [review]

File: diag_manager/fms_diag_file_object.F90 (hunk 2 above)
Comment (uramirez8707) [reply diag_manager/fms_diag_file_object.F90:956]: yes, it should be `_bnds`

File: diag_manager/fms_diag_file_object.F90 (hunk 3 above)
Comment (uramirez8707) [reply diag_manager/fms_diag_file_object.F90:1163]: this is doing io, which i think cannot be done in parallel

File: diag_manager/fms_diag_field_object.F90 [hunk 7]
```diff
@@ -788,6 +830,178 @@ pure function get_type_of_domain (this) &
...
+  if (this%is_static()) then
+    naxis = size(this%axis_ids)
+  else
+    naxis = size(this%axis_ids) + 1 !< Adding 1 more dimension for the unlimited dimension
+  endif
+
+  allocate(rslt(naxis))
+
+  do i = 1, size(this%axis_ids)
+    j = this%axis_ids(i)
```
[the remark this answers is not in the document]
Comment (uramirez8707) [reply diag_manager/fms_diag_field_object.F90]: This was replaced with a pointer https://github.com/NOAA-GFDL/FMS/pull/1102/commits/6d7341622dc912f07180a080e30fee9914b95f5e

File: diag_manager/fms_diag_file_object.F90 (hunk 4 above)
Comment (uramirez8707) [reply diag_manager/fms_diag_file_object.F90]: This was replaced with a pointer https://github.com/NOAA-GFDL/FMS/pull/1102/commits/6d7341622dc912f07180a080e30fee9914b95f5e

File: diag_manager/fms_diag_file_object.F90 (hunk 6 above)
Comment (thomas-robinson) [reply diag_manager/fms_diag_file_object.F90:1164]: The `where` won't work because you are writing.

**Review by thomas-robinson — APPROVED** [review]

**Pull request merged**